src/Npgsql/NpgsqlParameter.cs

```diff
@@ -506,60 +495,254 @@ public sealed override string SourceColumn
+                    return;
+                }
+
+                TypeInfo = options.GetDefaultTypeInfo(pgTypeId ?? options.ToCanonicalTypeId(options.UnknownPgType))
+                           ?? throw new NotSupportedException($"No configured writer was found for parameter with {
+                               (_npgsqlDbType is not null ? $"NpgsqlDbType '{_npgsqlDbType}'" : $" DataTypeName '{_dataTypeName}'")}.");
+            }
             else
-                ThrowInvalidOperationException();
+            {
+                TypeInfo = options.GetTypeInfo(valueType, pgTypeId)
+                           ?? throw new NotSupportedException($"No configured writer was found for parameter of type {valueType}{(_npgsqlDbType is not null
+                                   ? $" and NpgsqlDbType '{_npgsqlDbType}'" : pgTypeId is not null ? $" and DataTypeName '{_dataTypeName}'" : "")}.");
+            }
         }
 
-        void ThrowInvalidOperationException()
+        // This step isn't part of BindValue because we need to know the PgTypeId beforehand for things like SchemaOnly with null values.
+        // We never reuse resolutions for resolvers across executions as a mutable value itself may influence the result.
+        // TODO we could expose a property on a Converter/TypeInfo to indicate whether it's immutable, at that point we can reuse.
+        if (!previouslyBound || TypeInfo is PgResolverTypeInfo)
         {
-            var parameterName = !string.IsNullOrEmpty(ParameterName) ? ParameterName : $"${Collection?.IndexOf(this) + 1}";
-            ThrowHelper.ThrowInvalidOperationException($"Parameter '{parameterName}' must have either its NpgsqlDbType or its DataTypeName or its Value set");
+            ResetConverterResolution();
+            var resolution = ResolveConverter(TypeInfo!);
+            Converter = resolution.Converter;
+            PgTypeId = resolution.PgTypeId;
+        }
+
+        bool TryGetRepresentationalTypeId(string dataTypeName, out PgTypeId pgTypeId)
+        {
+            if (options.TypeCatalog.TryGetPostgresTypeByName(dataTypeName, out var pgType))
+            {
+                pgTypeId = options.ToCanonicalTypeId(pgType.GetRepresentationalType());
+                return true;
+            }
+
+            pgTypeId = default;
+            return false;
         }
     }
 
-    internal void Bind(TypeMapper typeMapper)
+    private protected virtual PgConverterResolution ResolveConverter(PgTypeInfo typeInfo) => typeInfo.GetObjectResolution(_value);
+
+    /// Bind the current value to the type info, truncate (if applicable), take its size, and do any final validation before writing.
+    internal void Bind(out DataFormat format, out Size size)
     {
-        ResolveHandler(typeMapper);
-        FormatCode = Handler!.PreferTextWrite ? FormatCode.Text : FormatCode.Binary;
+        if (TypeInfo is null)
```

Comment:
Not sure, but should this be an assertion?

Reply:
It's a null check to prevent later NREs due to not having called ResolveTypeInfo beforhand. It really won't be a performance crime.

Reply:
The comment was just about clarity, not perf, i.e. will this really happen in a legitimate user scenario etc.

Reply:
Right, not really, I have improved the exception message somewhat.